### PR TITLE
Istio MCS Generating ServiceExports

### PR DESCRIPTION
**Pull request by josephpeacock**
This PR accomplishes stage 2 in the [Istio MCS RFC](https://docs.google.com/document/u/2/d/1K8hvQ83UcJ9a7U8oqXIefwr6pFJn-VBEi40Ak-fwQtk/edit#heading=h.xw1gqgyqs5b)

In short, it allows Istio to be configured to create ServiceExport objects for non cluster-local services in an Istio-managed cluster.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by josephpeacock** [comment]
This PR accomplishes https://github.com/istio/istio/issues/29385 and is dependent upon https://github.com/istio/istio/pull/29763

File: manifests/charts/base/files/gen-istio-cluster.yaml [hunk 1]
```diff
@@ -3620,6 +3620,11 @@ rules:
   - apiGroups: [""]
     resources: ["secrets"]
     verbs: ["get", "watch", "list"]
+
+  #Used for MCS serviceexport management
+  - apiGroups: [""]
```
Comment (hzxuzhonghu) [inline manifests/charts/base/files/gen-istio-cluster.yaml:3625]: Is this in core group, i think it can not be.

File: manifests/charts/base/files/gen-istio-cluster.yaml [hunk 2]
```diff
@@ -3620,6 +3620,11 @@ rules:
...
+    resources: ["serviceexports"]
+    verbs: ["create", "delete"]
```
Comment (hzxuzhonghu) [inline manifests/charts/base/files/gen-istio-cluster.yaml:3627]: no `get` `list` needed?

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+)
+
+type ServiceExportController struct {
+	client        versioned.Interface
+	serviceClient corev1.CoreV1Interface
+
+	queue           queue.Instance
+	serviceInformer cache.SharedInformer
+
+	clusterLocalHosts []string //hosts marked as cluster-local, which will not have serviceeexports generated
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:42]: what is this for, concerned about the peformance

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+func NewServiceExportController(kubeClient kube.Client, clusterLocalHosts []string) (*ServiceExportController, error) {
+	serviceExportController := &ServiceExportController{
+		client:            kubeClient.MCSApis(),
+		serviceClient:     kubeClient.Kube().CoreV1(),
+		queue:             queue.NewQueue(time.Second),
+		clusterLocalHosts: clusterLocalHosts,
+	}
+
+	serviceExportController.serviceInformer = kubeClient.KubeInformer().Core().V1().Services().Informer()
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:53]: Donot create a new informer, this will duplicate the cache

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+		hostComponents := strings.Split(host, ".")
+		//name match
+		if (hostComponents[0] == "*" || hostComponents[0] == service.Name) && (hostComponents[1] == "*" || hostComponents[1] == service.Namespace) {
+			return true
+		}
+	}
+	return false
+}
+
+func (sc *ServiceExportController) createServiceExportIfNotPresent(service *v1.Service) error {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:124]: Who will consume the ServiceExports created here

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
From the implement, could not see when the serviceExport will be consumed and by who

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go (hunk 5 above)
Comment (josephpeacock) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:124]: The short answer is anyone will be able to. I'm working in parallel with @nmittler on the implementation of the [Istio MCS RFC](https://docs.google.com/document/d/1K8hvQ83UcJ9a7U8oqXIefwr6pFJn-VBEi40Ak-fwQtk/edit#heading=h.xw1gqgyqs5b). I took "phase 2", and there is another issue open for "phase 1" (https://github.com/istio/istio/issues/29384)

The consumption of ServiceExports is out of scope for this PR.

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:124]: It is not clear how this works from the RFC, should wait for the draft implement.

As i read through the mcs-api desgin, it is not clear to me how this can solve multicluster topology issue. 

Can you elaborate on how the EndpointSlice create is used?  https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#importing-services

File: manifests/charts/base/files/gen-istio-cluster.yaml (hunk 2 above)
Comment (howardjohn) [reply manifests/charts/base/files/gen-istio-cluster.yaml:3627]: and watch, probably update and patch will be needed as well?

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go [hunk 6]
```diff
@@ -79,6 +79,14 @@ type Multicluster struct {
 	secretNamespace  string
 	secretController *secretcontroller.Controller
 	syncInterval     time.Duration
+
+	//MCS Settings - settings for behavior of controllers relating to MCS
+	mcsSettings *MCSSettings
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:84]: nit: there is no need for this to be a pointer as far as I know

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go (hunk 3 above)
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:42]: I am concerned about the API here. Why are we mixing an Istio API ("cluster local") with the MCSD api? Isn't the goal of MCSD that we are just another implementor?

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go (hunk 4 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:53]: Its using shared index informer under the hood, it should be shared

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	serviceExport := v1alpha1.ServiceExport{}
+	serviceExport.Namespace = service.Namespace
+	serviceExport.Name = service.Name
+
+	_, err := sc.client.MulticlusterV1alpha1().ServiceExports(service.Namespace).Create(context.TODO(), &serviceExport, metav1.CreateOptions{})
+
+	if err != nil && strings.Contains(err.Error(), "already exists") {
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:131]: nit: use errors.IsAlreadyExists

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+		err = nil //This is the error thrown by the client if there is already an object with the name in the namespace. If that's true, we do nothing
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:132]: what if the contents are wrong?

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	err := sc.client.MulticlusterV1alpha1().ServiceExports(service.Namespace).Delete(context.TODO(), service.Name, metav1.DeleteOptions{})
+
+	if err != nil && strings.Contains(err.Error(), "not found") {
+		err = nil //If it's already gone, then we're happy
+	}
+	return err
+}
+
+func (sc *ServiceExportController) doInitialFullSync() {
+	allServices, err := sc.serviceClient.Services("").List(context.TODO(), metav1.ListOptions{})
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:148]: Calling .List() is not acceptable for performance, we should be using the informer cache. Direct API calls like this have bit us many times in the past

**Review by howardjohn — CHANGES_REQUESTED** [review]
I don't really understand why Istio is responsible for this. I presume MCSD will have multiple implementors, otherwise its not really useful. Will linkerd, consul, etc all be sending similar PRs, writing the same exact implementation of the same thing in 3 slightly different ways?

I would have expected the MCSD repo to have a controller, which is either deployed standalone, or *maybe*, once its past experimental, linked into Istiod as a library to reduce the deployment complexity.

I am a bit concerned that every project is going to implement the same controller slightly different and we will not actually have a consistent API like was desired.

**Comment by josephpeacock** [comment]
@howardjohn @hzxuzhonghu Maybe this is something we should discuss at the next Environments working group meeting? I was under the impression that the [approved RFC](https://docs.google.com/document/d/1K8hvQ83UcJ9a7U8oqXIefwr6pFJn-VBEi40Ak-fwQtk/edit#heading=h.xw1gqgyqs5b) represented an agreement that we were going to build the specified functionality into Istio. In meeting with @nmittler, we decided that phases one and two from that document could be done in parallel, and that I would put together a PR encompassing only stage two. 

I'm happy to shelve or permanently close this PR if it's agreed that building MCS functionality into Istio is the wrong direction for the project. You both had some very good specific feedback in the code as well, and I appreciate the prompt review. But I'd like for us to agree on the big picture direction before I spend more time rebasing and/or updating this PR.

**Comment by howardjohn** [comment]
To clarify a bit. I am not *against* this, I am just not sure it's the
right thing to do without more info (answers to questions above about the
wider ecosystem). The RFC is a bit light on the details. Discussion in env
WG (or this PR) SGTM! Thanks for working on this

**Comment by josephpeacock** [comment]
Fair enough, let me answer the questions that I feel that I can speak to:

> I don't really understand why Istio is responsible for this. I presume MCSD will have multiple implementors, otherwise its not really useful. Will linkerd, consul, etc all be sending similar PRs, writing the same exact implementation of the same thing in 3 slightly different ways?

I think this is the intention of the [k8s spec.](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api) They explicitly say in their document that anything should be able to be an MCS controller:

```
mcs-controller - A controller that syncs services across clusters and makes them available for multi-cluster service discovery and connectivity. There may be multiple implementations, this doc describes expected common behavior. The controller may be a single controller, multiple decentralized controllers, or a human using kubectl to create resources. This document aims to support any implementation that fulfills the behavioral expectations of this API.

```

In this case, I'd argue it's a question of value-add. Do we see value in building out MCS controller functionality in Istio? (with the alternative being having users need to use another MCS controller in tandem with Istio if they want that functionality). And I do think being able to use service names to route across multiple clusters would really simplify some multicluster Istio installations.

> I would have expected the MCSD repo to have a controller, which is either deployed standalone, or maybe, once its past experimental, linked into Istiod as a library to reduce the deployment complexity.

The only official k8s MCS repo that I know of at the moment is for the [MCS API](https://github.com/kubernetes-sigs/mcs-api). I don't know if there is a plan for an official k8s MCS controller, or if they are intending to leave it entirely up to the community.

> I am a bit concerned that every project is going to implement the same controller slightly different and we will not actually have a consistent API like was desired.

You're probably right that there may be opportunities for a wider conversation about how specific the API is and should be moving forward on the k8s side. But I'm not sure if that's relevant to Istio's implementation, unless we want to wait to implement until the spec is more well-defined.

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go (hunk 3 above)
Comment (josephpeacock) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:42]: > Isn't the goal of MCSD that we are just another implementor?

Yes, but one implementation detail that we get to determine is which services are exported (the API intentionally leaves that up to whatever entity is generating serviceExports). So we're piggybacking off of this Istio API with the assumption that cluster local hosts should not be exported for MCS.

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go (hunk 8 above)
Comment (josephpeacock) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:132]: For this implementation, we are making the intentional choice that we don't care, and don't support cases in which some other entity (either human or 3rd party MCS controller) modifies a ServiceExport. So if it exists, we stop processing (to minimize cases where the other entity overwrite may be desired behavior).

File: pilot/pkg/bootstrap/options.go [hunk 10]
```diff
@@ -45,6 +45,10 @@ type RegistryOptions struct {
 
 	// DistributionTracking control
 	DistributionTrackingEnabled bool
+
+	MCSServiceExportEnabled bool
+
+	ClusterLocalHosts []string
```
Comment (nmittler) [inline pilot/pkg/bootstrap/options.go:51]: Doesn't this duplicate the configuration in MeshConfig?

File: pilot/pkg/bootstrap/server.go [hunk 11]
```diff
@@ -232,6 +232,11 @@ func NewServer(args *PilotArgs) (*Server, error) {
 	s.initMeshNetworks(args, s.fileWatcher)
 	s.initMeshHandlers()
 
+	e.PushContext.Mesh = e.Mesh()
+	args.RegistryOptions.ClusterLocalHosts = e.PushContext.GetClusterLocalHosts(e)
```
Comment (nmittler) [inline pilot/pkg/bootstrap/server.go:236]: Can't meshconfig be updated?

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go (hunk 5 above)
Comment (nmittler) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:124]: @hzxuzhonghu 

The basic MCS flow is:

1. Something generates a `ServiceExport` for a service in a cluster
2. The MCS controller (Istio?) sees the `ServiceExport` and creates `ServiceImport` and `EndpointSlice`s in all of the *other* clusters in the clusterset (i.e. mesh).
3. In the other clusters, the `ServiceImport` and `EndpointSlice`s are used by K8s to program kube proxy to allow service calls to span clusters.

What this PR is attempting to accomplish is 1. This simply creates the `ServiceExport` resource ... it does nothing with `EndpointSlice`s. This will have no real impact on the system if there is no MCS controller performing the `ServiceImport`.

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+		return nil //don't don anything for marked clusterlocal services
+	}
+	return sc.createServiceExportIfNotPresent(obj)
+}
+
+func (sc *ServiceExportController) HandleDeletedService(obj *v1.Service) error {
+	return sc.deleteServiceExportIfPresent(obj)
+}
+
+func convertToService(obj interface{}) (*v1.Service, error) {
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:98]: We do this elsewhere ... can we use a function?

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+		}
+		cm, ok = tombstone.Obj.(*v1.Service)
+		if !ok {
+			return nil, fmt.Errorf("tombstone contained object that is not a Service %#v", obj)
+		}
+	}
+	return cm, nil
+}
+
+func (sc *ServiceExportController) isServiceClusterLocal(service *v1.Service) bool {
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:113]: Can we move this function to the context?

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+		err = nil //This is the error thrown by the client if there is already an object with the name in the namespace. If that's true, we do nothing
+	}
+	return err
+}
+
+func (sc *ServiceExportController) deleteServiceExportIfPresent(service *v1.Service) error {
+	//cannot use the auto-generated client as it hardcodes the namespace in the client struct, and we can't have one client per watched ns
+	err := sc.client.MulticlusterV1alpha1().ServiceExports(service.Namespace).Delete(context.TODO(), service.Name, metav1.DeleteOptions{})
+
+	if err != nil && strings.Contains(err.Error(), "not found") {
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:141]: Is there a standard error here, similar to errors.IsAlreadyExists?

File: manifests/charts/base/files/gen-istio-cluster.yaml (hunk 2 above)
Comment (josephpeacock) [reply manifests/charts/base/files/gen-istio-cluster.yaml:3627]: This PR only needs create/delete on serviceexports. The eventual PR to consume them will likely need the ability to watch, but I want to add the minimal set of permissions necessary for this functionality.

File: manifests/charts/base/files/gen-istio-cluster.yaml (hunk 1 above)
Comment (josephpeacock) [reply manifests/charts/base/files/gen-istio-cluster.yaml:3625]: I'm a bit unfamiliar with the manifests, could you recommend a more appropriate place?

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go (hunk 13 above)
Comment (josephpeacock) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:102]: I'm not sure what you mean by this, can you clarify?

**Comment by josephpeacock** [comment]
Ready for re-review.

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go (hunk 13 above)
Comment (nmittler) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:102]: The context is where you're getting information regarding cluster-local services. Could it just expose this method instead of defining it here?

**Comment by nmittler** [comment]
@josephpeacock my main concern still hasn't been addressed. You're providing a copy of the cluster-local service list at construction time, which means it will be constant for the life of istiod. However, MeshConfig can be changed dynamically (including this list). My suggestion was to always go to the push context to determine whether or not a service is cluster-local, rather than using a copy of the list which may be outdated.

**Comment by josephpeacock** [comment]
Ahh, I definitely misunderstood there. I was actively trying to avoid using the push context in my controller because I had thought it was specifically for pushing metrics and errors (as per the comments on the struct). But your point makes a lot of sense, I've shifted to that approach.

File: pilot/pkg/bootstrap/servicecontroller.go [hunk 15]
```diff
@@ -81,6 +81,11 @@ func (s *Server) initKubeRegistry(args *PilotArgs) (err error) {
 		caBundlePath = args.ServerOptions.TLSOptions.CaCertFile
 	}
 
+	mcsSettings := kubecontroller.MCSSettings{
+		EnableServiceExport: false, //todo change to s.environment.Mesh().EnableMcsServiceExport once https://github.com/istio/api/pull/1785 is merged,
+		PushContext:         s.environment.PushContext,
```
Comment (nmittler) [inline pilot/pkg/bootstrap/servicecontroller.go:86]: I think what you want is the environment, not the pushcontext, actually. The pushcontext is set as a property of the environment, but it can be overwritten and is therefore not stable for concurrent access. Environment should be stable throughout the life of the process.

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 16]
```diff
@@ -89,7 +89,7 @@ func (sc *ServiceExportController) Run(stopCh <-chan struct{}) {
 }
 
 func (sc *ServiceExportController) HandleNewService(obj *v1.Service) error {
-	if sc.pushContext.IsServiceClusterLocal(obj) {
+	if sc.environment.PushContext.IsServiceClusterLocal(obj) {
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:92]: You don't want to use the `PushContext` here ... instead do something like `sc.environment.Mesh().IsServiceClusterLocal(obj)`, or maybe `sc.environment.IsServiceClusterLocal(obj)`

File: pilot/pkg/model/push_context.go [hunk 17]
```diff
@@ -1663,54 +1660,6 @@ func (ps *PushContext) initMeshNetworks() {
 
 }
 
-func (ps *PushContext) initClusterLocalHosts(e *Environment) {
```
[the remark this answers is not in the document]
Comment (josephpeacock) [reply pilot/pkg/model/push_context.go:1666]: I thought about that, but it would require a more substantial refactor than (in my opinion) we want to build into this PR. `PushContext.IsClusterLocal` is used by both the [cluster_builder](https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/cluster_builder.go#L304) and [endpoint_builder](https://github.com/istio/istio/blob/master/pilot/pkg/xds/endpoint_builder.go#L242) and, as far as I can tell, neither of those have a trivial way to leverage the environment for this information.

I'm happy to do a bigger refactor here if you think it's the appropriate way to move forward, but that was my rationale for duplicating the clusterLocalHosts info in both the push context and the environment.

[the remark this answers is not in the document]
Comment (nmittler) [reply pilot/pkg/model/push_context.go:1666]: Understood. Could we at least move the population of `clusterLocalHosts` to the environment and have it copied over when we create the push context? I think that should work.

[the remark this answers is not in the document]
Comment (josephpeacock) [reply pilot/pkg/model/push_context.go:1666]: I believe that's what I have done [here](https://github.com/istio/istio/pull/29763/files#diff-a8e332e06b003dfbf58c29f37de893fbb0e45f642200ea3ea2faa9847fc7e1e0R915) - when `pushContext.InitContext` is called, it calls out to the environment to generate the cluster local hosts, which updates the env list as well as the one in the newly initialized pushcontext

[the remark this answers is not in the document]
Comment (nmittler) [reply pilot/pkg/model/push_context.go:1666]: Ah you're right. The part that's missing is the bit about `initMeshHandlers`. We don't want to re-initialize `clusterLocalHosts` each time we create a `PushContext`. Instead, we only want to re-initialize it when the MeshConfig changes. To do that, we need to add logic to `initMeshHandlers` to call `env.initClusterLocalHosts`. Then, when we create a `PushContext`, we just copy over the value from the `Environment`.

File: pilot/pkg/model/context.go [hunk 18]
```diff
@@ -140,6 +143,65 @@ func (e *Environment) Version() string {
...
+		}
+	}
+
+	sort.Sort(host.Names(clusterLocalHosts))
+	e.clusterLocalHosts = clusterLocalHosts
+	return clusterLocalHosts
+}
+
+func (e *Environment) GetClusterLocalHosts() []host.Name {
+	return e.clusterLocalHosts
```
Comment (nmittler) [inline pilot/pkg/model/context.go:196]: Do we need to lock around this?

File: pilot/pkg/bootstrap/server.go [hunk 19]
```diff
@@ -1125,4 +1125,6 @@ func (s *Server) initMeshHandlers() {
 			Reason: []model.TriggerReason{model.GlobalUpdate},
 		})
 	})
+
+	s.environment.InitClusterLocalHosts()
```
Comment (nmittler) [inline pilot/pkg/bootstrap/server.go:1129]: This needs to be moved to the `AddMeshHandler` function above ... before we do anything with the `XDSServer`.

File: pilot/pkg/model/context.go (hunk 18 above)
Comment (josephpeacock) [reply pilot/pkg/model/context.go:196]: It's only used during the initialization of a PushContext. So the potential race condition would be if the meshconfig was updated, and while the `InitClusterLocalHosts` was processing, a new PushContext could be created with the stale set of clusterLocalHosts.

That's definitely a race condition that also exists in the current code, but that shouldn't stop us from fixing it if necessary.

I'll defer to your understanding of how the push context is used, but my gut would say that's a small enough hole with small enough consequences that we don't need to worry about it, as I believe push contexts are meant to be relatively ephemeral anyways (and Istio follows a general consistency model) so whenever the next push is done, things will come back in line with the updated config.

Comment (nmittler) [reply pilot/pkg/model/context.go:196]: @howardjohn any thoughts around the safety of this?

Comment (nmittler) [reply pilot/pkg/model/context.go:196]: @josephpeacock perhaps we should just add a lock here to be on the safe side? My concern is that we're adding (additional?) asynchronous access since your controller is going to be accessing it as well.

Comment (josephpeacock) [reply pilot/pkg/model/context.go:196]: I'm definitely fine adding it. Do you think we also want it around the `InitClusterLocalHosts`?

Comment (nmittler) [reply pilot/pkg/model/context.go:200]: yeah, just lock before we actually set it.

**Comment by josephpeacock** [comment]
@nmittler I got confirmation from @howardjohn that there is currently no pattern for including 3rd party CRDs in the Istio manifests, so we need to have users install them manually (usually included as part of documentation).

**Comment by nmittler** [comment]
@josephpeacock SGTM, thanks!

File: tests/integration/pilot/mcs/serviceexport_test.go [hunk 20]
```diff
@@ -89,6 +90,38 @@ func TestServiceExports(t *testing.T) {
 		Run(func(ctx framework.TestContext) {
 			cluster := ctx.Clusters().Default()
 
+			// restarting the istiod deployment because we short-circuit the serviceexport controller if the CRD isn't there on istiod startup
```
Comment (nmittler) [inline tests/integration/pilot/mcs/serviceexport_test.go:93]: Can you clarify why this is needed? This sounds like a bug

Comment (josephpeacock) [reply tests/integration/pilot/mcs/serviceexport_test.go:93]: That goes back to this change: https://github.com/istio/istio/pull/29763/files#diff-dabdc9b878f95aed9dee67d29e869812e924607537416f2aa9ebf4f79ed0142fR96 which I think we discussed in one of the Intuit-MCS meetings. Essentially - we don't have a way for the controller to start or stop itself so we just permanently short circuit the logic if the CRD doesn't exist, and log a message telling the user to apply the CRD and restart istiod.

We could do some form of periodic retries if we wanted to, but that also seems like a very edge case

Comment (howardjohn) [reply tests/integration/pilot/mcs/serviceexport_test.go:93]: can you follow the pattern used here: https://github.com/istio/istio/blob/fd6ecf7e4edbb080c31a0146c856c30ae66efb87/tests/integration/pilot/main_test.go#L55 so we don't need to restart?

File: manifests/charts/base/templates/crds.yaml [hunk 21]
```diff
@@ -1,4 +1,4 @@
 {{- if .Values.base.enableCRDTemplates }}
 {{ .Files.Get "crds/crd-all.gen.yaml" }}
 {{ .Files.Get "crds/crd-operator.yaml" }}
-{{- end }}
+{{- end }}
```
Comment (nmittler) [inline manifests/charts/base/templates/crds.yaml:4]: nit: revert this change?

File: pilot/pkg/model/context.go [hunk 22]
```diff
@@ -146,6 +151,77 @@ func (e *Environment) Version() string {
...
+}
+
+func (e *Environment) GetClusterLocalHosts() []host.Name {
+	e.mutex.Lock()
+	defer e.mutex.Unlock()
+	return e.clusterLocalHosts
+}
+
+func (e *Environment) IsServiceClusterLocal(service *v1.Service) bool {
+	hostname := fmt.Sprintf("%s.%s.svc.cluster.local", service.Name, service.Namespace)
```
Comment (nmittler) [inline pilot/pkg/model/context.go:220]: I think we need to lock everywhere we access `e.clusterLocalHosts`.

File: pilot/pkg/model/context.go [hunk 23]
```diff
@@ -154,6 +230,14 @@ func (e *Environment) SetLedger(l ledger.Ledger) {
 	e.ledger = l
 }
 
+// only user for testing
+func (e *Environment) SetClusterLocalHosts(hostList []string) {
+	e.clusterLocalHosts = make([]host.Name, len(hostList))
```
Comment (nmittler) [inline pilot/pkg/model/context.go:235]: And here ... maybe you should create the list first and then just lock when you assign it to `e.clusterLocalHosts`.

File: pilot/pkg/model/context.go (hunk 22 above)
Comment (nmittler) [reply pilot/pkg/model/context.go:220]: nit: could you just unlock before the return rather than using `defer`? There is a slight performance penalty for using defer.

File: pilot/pkg/model/context.go [hunk 24]
```diff
@@ -154,6 +232,17 @@ func (e *Environment) SetLedger(l ledger.Ledger) {
...
 }
 
+// only user for testing
+func (e *Environment) SetClusterLocalHosts(hostList []string) {
+	clusterLocalHosts := make([]host.Name, len(hostList))
+	for _, hostName := range hostList {
+		clusterLocalHosts = append(clusterLocalHosts, host.Name(hostName))
+	}
+	e.mutex.Lock()
+	defer e.mutex.Lock()
```
Comment (nmittler) [inline pilot/pkg/model/context.go:242]: same here.

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 25]
```diff
@@ -0,0 +1,142 @@
...
+	"sigs.k8s.io/mcs-api/pkg/client/clientset/versioned"
+
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/queue"
+	"istio.io/pkg/log"
+)
+
+type ServiceExportController struct {
+	client        versioned.Interface
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:36]: @howardjohn do you think we should just add these to the kube client API?

Comment (nmittler) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:36]: oh nm ... it's already there :)

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+func (sc *ServiceExportController) Run(stopCh <-chan struct{}) {
+	cache.WaitForCacheSync(stopCh, sc.serviceInformer.HasSynced)
+	log.Infof("Syncing existing services and serviceexports...")
+	sc.doInitialFullSync()
+	log.Infof("serviceexport sync complete")
+	log.Infof("ServiceExport controller started")
+	go sc.queue.Run(stopCh)
+}
+
+func (sc *ServiceExportController) HandleNewService(obj *v1.Service) error {
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:95]: Why are we exporting this method?

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+	if !sc.crdExists {
+		return nil
+	}
+	if sc.environment.IsServiceClusterLocal(obj) {
+		return nil // don't do anything for marked clusterlocal services
+	}
+	return sc.createServiceExportIfNotPresent(obj)
+}
+
+func (sc *ServiceExportController) HandleDeletedService(obj *v1.Service) error {
```
Comment (nmittler) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:105]: same question here

File: pkg/kube/client.go [hunk 28]
```diff
@@ -346,6 +361,12 @@ func newClientInternal(clientFactory util.Factory, revision string) (*client, er
 	}
 	c.gatewayapiInformer = gatewayapiinformer.NewSharedInformerFactory(c.gatewayapi, resyncInterval)
 
+	c.mcsapis, err = mcsapisClient.NewForConfig(c.config)
+	if err != nil {
+		return nil, err
+	}
+	c.mcsapisInformers = mcsapisInformer.NewSharedInformerFactory(c.mcsapis, resyncInterval)
```
Comment (nmittler) [inline pkg/kube/client.go:368]: This is done regardless of the environment variable. Will there be any impact on Istio when disabled?

File: tests/integration/pilot/mcs/serviceexport_test.go (hunk 20 above)
Comment (nmittler) [reply tests/integration/pilot/mcs/serviceexport_test.go:92]: Alternatively you could install it on the clusters in the prow script before running `go test`.

File: pilot/pkg/model/context.go (hunk 24 above)
Comment (nmittler) [reply pilot/pkg/model/context.go:242]: oh sorry ... that as a slight perf improvement, you should manually unlock at the end rather than using `defer`.

Comment (josephpeacock) [reply pilot/pkg/model/context.go:242]: Hmm, TIL. I thought that was the preferred way of handling unlocks in Go - will fix it.

Comment (nmittler) [reply pilot/pkg/model/context.go:242]: It is, if the function is long and/or has multiple returns. The downside is that there's a slight performance cost. So if it's simple or if the method is called frequently, prefer to explicitly unlock.

File: pkg/kube/client.go (hunk 28 above)
Comment (josephpeacock) [reply pkg/kube/client.go:368]: That's a good point - as per https://github.com/kubernetes/kubernetes/issues/79610 it at least will flood kube-controller logs which isn't great. 

I don't think my solution is the cleanest, so I'd welcome alternative suggestions. I just had no idea how to get the env variable value into the client creation other than checking it in the method itself

Comment (josephpeacock) [reply pkg/kube/client.go]: Awesome. I'll undo that change

File: pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+func (sc *ServiceExportController) handleDeletedService(obj *v1.Service) error {
+	return sc.deleteServiceExportIfPresent(obj)
+}
+
+func (sc *ServiceExportController) createServiceExportIfNotPresent(service *v1.Service) error {
+	serviceExport := v1alpha1.ServiceExport{}
+	serviceExport.Namespace = service.Namespace
+	serviceExport.Name = service.Name
```
Comment (rcernich) [inline pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:112]: Do you want to add an owner reference to this?  This would also eliminate the need for handling deletions.

Comment (josephpeacock) [reply pilot/pkg/serviceregistry/kube/controller/serviceexportcontroller.go:112]: That is actually something that is explicitly called out by the [K8s MCS Spec](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#objectreference-in-serviceexportspec-to-directly-map-to-a-service), and was decided against. Instead, they chose to rely entirely on the name mapping. I tried to make this implementation match the spec as closely as possible.

**Comment by nmittler** [comment]
@josephpeacock thanks for all the help. I'm taking this over in #31401. Shall we just close this one out?

**Comment by josephpeacock** [comment]
Yup, let's do that. Thanks for taking this over!
